File: test/Models/pheno_pkg/src/simplace/Vernalizationprogress.java
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import javafx.util.*;
import java.time.LocalDateTime;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom.Element;


public class Vernalizationprogress extends FWSimComponent
{
    private FWSimVariable<Double> minTvern;
    private FWSimVariable<Double> intTvern;
    private FWSimVariable<Double> vAI;
    private FWSimVariable<Double> vBEE;
    private FWSimVariable<Double> minDL;
    private FWSimVariable<Double> maxDL;
    private FWSimVariable<Double> maxTvern;
    private FWSimVariable<Double> pNini;
    private FWSimVariable<Double> aMXLFNO;
    private FWSimVariable<Integer> isVernalizable;
    private FWSimVariable<Double> dayLength;
    private FWSimVariable<Double> deltaTT;
    private FWSimVariable<Double> cumulTT;
    private FWSimVariable<LocalDateTime> currentdate;
    private FWSimVariable<Double> leafNumber_t1;
    private FWSimVariable<String[]> calendarMoments_t1;
    private FWSimVariable<LocalDateTime[]> calendarDates_t1;
    private FWSimVariable<Double[]> calendarCumuls_t1;
    private FWSimVariable<Double> vernaprog_t1;
    private FWSimVariable<Double> minFinalNumber_t1;
    private FWSimVariable<Double> vernaprog;
    private FWSimVariable<Double> minFinalNumber;
    private FWSimVariable<String[]> calendarMoments;
    private FWSimVariable<LocalDateTime[]> calendarDates;
    private FWSimVariable<Double[]> calendarCumuls;

    public Vernalizationprogress(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Vernalizationprogress(){
        super();
    }
    @Override
    protected void process()
    {
        double tdayLength = dayLength.getValue();
        double tdeltaTT = deltaTT.getValue();
        double tcumulTT = cumulTT.getValue();
        double tleafNumber_t1 = leafNumber_t1.getValue();
        List<String> tcalendarMoments_t1 = Arrays.asList(calendarMoments_t1.getValue());
        List<LocalDateTime> tcalendarDates_t1 = Arrays.asList(calendarDates_t1.getValue());
        List<Double> tcalendarCumuls_t1 = Arrays.asList(calendarCumuls_t1.getValue());
        double tminTvern = minTvern.getValue();
        double tintTvern = intTvern.getValue();
        double tvAI = vAI.getValue();
        double tvBEE = vBEE.getValue();
        double tminDL = minDL.getValue();
        double tmaxDL = maxDL.getValue();
        double tmaxTvern = maxTvern.getValue();
        double tpNini = pNini.getValue();
        double taMXLFNO = aMXLFNO.getValue();
        double tvernaprog_t1 = vernaprog_t1.getValue();
        LocalDateTime tcurrentdate = currentdate.getValue();
        int tisVernalizable = isVernalizable.getValue();
        double tminFinalNumber_t1 = minFinalNumber_t1.getValue();
        double tvernaprog;
        double tminFinalNumber;
        List<String> tcalendarMoments;
        List<LocalDateTime> tcalendarDates;
        List<Double> tcalendarCumuls;
        double maxVernaProg;
        double dLverna;
        double primordno;
        double minLeafNumber;
        double potlfno;
        double tt;
        tcalendarMoments = new ArrayList<>(tcalendarMoments_t1);
        tcalendarCumuls = new ArrayList<>(tcalendarCumuls_t1);
        tcalendarDates = new ArrayList<>(tcalendarDates_t1);
        tminFinalNumber = tminFinalNumber_t1;
        tvernaprog = tvernaprog_t1;
        if (tisVernalizable == 1 && tvernaprog_t1 < 1.0d)
        {
            tt = tdeltaTT;
            if (tt >= tminTvern && tt <= tintTvern)
            {
                tvernaprog = tvernaprog_t1 + (tvAI * tt) + tvBEE;
            }
            else
            {
                tvernaprog = tvernaprog_t1;
            }
            if (tt > tintTvern)
            {
                maxVernaProg = tvAI * tintTvern + tvBEE;
                dLverna = Math.max(tminDL, Math.min(tmaxDL, tdayLength));
                tvernaprog = tvernaprog + Math.max(0.0d, maxVernaProg * (1.0d + ((tintTvern - tt) / (tmaxTvern - tintTvern) * ((dLverna - tminDL) / (tmaxDL - tminDL)))));
            }
            primordno = 2.0d * tleafNumber_t1 + tpNini;
            minLeafNumber = tminFinalNumber_t1;
            if (tvernaprog >= 1.0d || primordno >= taMXLFNO)
            {
                tminFinalNumber = Math.max(primordno, tminFinalNumber_t1);
                tcalendarMoments.add("EndVernalisation");
                tcalendarCumuls.add(tcumulTT);
                tcalendarDates.add(tcurrentdate);
                tvernaprog = Math.max(1.0d, tvernaprog);
            }
            else
            {
                potlfno = taMXLFNO - ((taMXLFNO - minLeafNumber) * tvernaprog);
                if (primordno >= potlfno)
                {
                    tminFinalNumber = Math.max((potlfno + primordno) / 2.0d, tminFinalNumber_t1);
                    tvernaprog = Math.max(1.0d, tvernaprog);
                    tcalendarMoments.add("EndVernalisation");
                    tcalendarCumuls.add(tcumulTT);
                    tcalendarDates.add(tcurrentdate);
                }
            }
        }
        vernaprog.setValue(tvernaprog, this);
        minFinalNumber.setValue(tminFinalNumber, this);
        calendarMoments.setValue(tcalendarMoments.toArray(new String[0]), this);
        calendarDates.setValue(tcalendarDates.toArray(new LocalDateTime[0]), this);
        calendarCumuls.setValue(tcalendarCumuls.toArray(new Double[0]), this);
    }

    @Override
    protected void init()
    {
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Vernalizationprogress(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("cminTvern", "Minimum temperature for vernalization to occur", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", -20, 60, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("cintTvern", "Intermediate temperature for vernalization to occur", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", -20, 60,  11.0, this));
        addVariable(FWSimVariable.createSimVariable("cvAI", "Response of vernalization rate to temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 1,  0.015, this));
        addVariable(FWSimVariable.createSimVariable("cvBEE", "Vernalization rate at 0°C", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 1, 0.01, this));
        addVariable(FWSimVariable.createSimVariable("cminDL", "Threshold daylength below which it does influence vernalization rate", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 12, 24, 8.0, this));
        addVariable(FWSimVariable.createSimVariable("cmaxDL", "Saturating photoperiod above which final leaf number is not influenced by daylength", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 24, 15.0, this));
        addVariable(FWSimVariable.createSimVariable("cmaxTvern", "Maximum temperature for vernalization to occur", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", -20, 60,  23.0, this));
        addVariable(FWSimVariable.createSimVariable("cpNini", "Number of primorida in the apex at emergence", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 24, 4.0, this));
        addVariable(FWSimVariable.createSimVariable("caMXLFNO", "Absolute maximum leaf number", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 25, 24.0, this));
        addVariable(FWSimVariable.createSimVariable("cisVernalizable", "true if the plant is vernalizable", DATA_TYPE.INT, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 1, 1, this));
        addVariable(FWSimVariable.createSimVariable("idayLength", "day length", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 10000, 12.3037621834005, this));
        addVariable(FWSimVariable.createSimVariable("ideltaTT", "difference cumul TT between j and j-1 day ", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", -20, 100, 20.3429985011972, this));
        addVariable(FWSimVariable.createSimVariable("icumulTT", "cumul thermal times at currentdate", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", -200, 10000, 112.330110409888, this));
        addVariable(FWSimVariable.createSimVariable("icurrentdate", "current date ", DATA_TYPE.DATE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, null, this));
        addVariable(FWSimVariable.createSimVariable("sleafNumber", "Actual number of phytomers", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 25, 0, this));
        addVariable(FWSimVariable.createSimVariable("scalendarMoments", "List containing appearance of each stage", DATA_TYPE.CHARARRAY, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, new String[] {"Sowing"}, this));
        addVariable(FWSimVariable.createSimVariable("scalendarDates", "List containing  the dates of the wheat developmental phases", DATA_TYPE.CHARARRAY, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, null, this));
        addVariable(FWSimVariable.createSimVariable("scalendarCumuls", "list containing for each stage occured its cumulated thermal times", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, new Double[] {0.0}, this));
        addVariable(FWSimVariable.createSimVariable("svernaprog", "progression on a 0  to 1 scale of the vernalization", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 1,  0.5517254187376879, this));
        addVariable(FWSimVariable.createSimVariable("sminFinalNumber", "minimum final leaf number", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0, 25, 5.5, this));

        return iFieldMap;
    }
}

File: test/Models/pheno_pkg/src/record/Vernalizationprogress.cpp
// @@tagdynamic@@
// @@tagdepends: vle.discrete-time @@endtagdepends

#include <vle/DiscreteTime.hpp>
namespace vd = vle::devs;
namespace vv = vle::value;
// Definition du namespace de la classe du modele
namespace record {
namespace Phenology {
using namespace vle::discrete_time;
class Vernalizationprogress: public DiscreteTimeDyn {
public:
    Vernalizationprogress(const vd::DynamicsInit& atom, const vd::InitEventList& events) : DiscreteTimeDyn(atom, events)
    {
        // Ces parametres ont une valeur par defaut utilise si la condition n'est pas definie
        minTvern = (events.exist("minTvern")) ? vv::toDouble(events.get("minTvern")) : 0.0;
        intTvern = (events.exist("intTvern")) ? vv::toDouble(events.get("intTvern")) :  11.0;
        vAI = (events.exist("vAI")) ? vv::toDouble(events.get("vAI")) :  0.015;
        vBEE = (events.exist("vBEE")) ? vv::toDouble(events.get("vBEE")) : 0.01;
        minDL = (events.exist("minDL")) ? vv::toDouble(events.get("minDL")) : 8.0;
        maxDL = (events.exist("maxDL")) ? vv::toDouble(events.get("maxDL")) : 15.0;
        maxTvern = (events.exist("maxTvern")) ? vv::toDouble(events.get("maxTvern")) :  23.0;
        pNini = (events.exist("pNini")) ? vv::toDouble(events.get("pNini")) : 4.0;
        aMXLFNO = (events.exist("aMXLFNO")) ? vv::toDouble(events.get("aMXLFNO")) : 24.0;
        isVernalizable = (events.exist("isVernalizable")) ? vv::toInteger(events.get("isVernalizable")) : 1;
        //  Variables gérées par ce composant
        vernaprog.init(this,"vernaprog", events);
        minFinalNumber.init(this,"minFinalNumber", events);
        calendarMoments.init(this,"calendarMoments", events);
        calendarDates.init(this,"calendarDates", events);
        calendarCumuls.init(this,"calendarCumuls", events);
        //  Variables gérées par un autre composant
        leafNumber.init(this,"leafNumber", events);
        dayLength.init(this,"dayLength", events);
        deltaTT.init(this,"deltaTT", events);
        cumulTT.init(this,"cumulTT", events);
        currentdate.init(this,"currentdate", events);
    }
    /**
    * @brief Destructeur de la classe du modèle.
    **/
    virtual ~Vernalizationprogress() {};
    /**
    * @brief Methode de calcul effectuée à chaque pas de temps
    * @param time la date du pas de temps courant
    */
    virtual void compute(const vd::Time& /*time*/)
    {
        double maxVernaProg;
        double dLverna;
        double primordno;
        double minLeafNumber;
        double potlfno;
        double tt;
        calendarMoments = calendarMoments_t1;
        calendarCumuls = calendarCumuls_t1;
        calendarDates = calendarDates_t1;
        minFinalNumber = minFinalNumber(-1);
        vernaprog = vernaprog(-1);
        if (isVernalizable == 1 && vernaprog(-1) < 1.0d)
        {
            tt = deltaTT();
            if (tt >= minTvern && tt <= intTvern)
            {
                vernaprog = vernaprog(-1) + (vAI * tt) + vBEE;
            }
            else
            {
                vernaprog = vernaprog(-1);
            }
            if (tt > intTvern)
            {
                maxVernaProg = vAI * intTvern + vBEE;
                dLverna = max(minDL, min(maxDL, dayLength()));
                vernaprog = vernaprog() + max(0.0d, maxVernaProg * (1.0d + ((intTvern - tt) / (maxTvern - intTvern) * ((dLverna - minDL) / (maxDL - minDL)))));
            }
            primordno = 2.0d * leafNumber(-1) + pNini;
            minLeafNumber = minFinalNumber(-1);
            if (vernaprog() >= 1.0d || primordno >= aMXLFNO)
            {
                minFinalNumber = max(primordno, minFinalNumber(-1));
                calendarMoments().push_back("EndVernalisation");
                calendarCumuls().push_back(cumulTT());
                calendarDates().push_back(currentdate());
                vernaprog = max(1.0d, vernaprog());
            }
            else
            {
                potlfno = aMXLFNO - ((aMXLFNO - minLeafNumber) * vernaprog());
                if (primordno >= potlfno)
                {
                    minFinalNumber = max((potlfno + primordno) / 2.0d, minFinalNumber(-1));
                    vernaprog = max(1.0d, vernaprog());
                    calendarMoments().push_back("EndVernalisation");
                    calendarCumuls().push_back(cumulTT());
                    calendarDates().push_back(currentdate());
                }
            }
        }
    }
private:
    //Variables d'etat
    /**
    * @brief progression on a 0  to 1 scale of the vernalization ()
    */
    Var vernaprog/**
    * @brief minimum final leaf number (leaf)
    */
    Var minFinalNumber/**
    * @brief List containing appearance of each stage ()
    */
    Vect calendarMoments/**
    * @brief List containing  the dates of the wheat developmental phases ()
    */
    Vect calendarDates/**
    * @brief list containing for each stage occured its cumulated thermal times ()
    */
    Vect calendarCumuls

    //Entrées
    /**
        * @brief Actual number of phytomers (leaf)
        */
    Var leafNumber/**
        * @brief day length (mm2 m-2)
        */
    Var dayLength/**
        * @brief difference cumul TT between j and j-1 day  (°C d)
        */
    Var deltaTT/**
        * @brief cumul thermal times at currentdate (°C d)
        */
    Var cumulTT/**
        * @brief current date  ()
        */
    Var currentdate

    //Paramètres du modele
    /**
    * @brief Minimum temperature for vernalization to occur (°C)
    */
    double minTvern;
    /**
    * @brief Intermediate temperature for vernalization to occur (°C)
    */
    double intTvern;
    /**
    * @brief Response of vernalization rate to temperature (d-1 °C-1)
    */
    double vAI;
    /**
    * @brief Vernalization rate at 0°C (d-1)
    */
    double vBEE;
    /**
    * @brief Threshold daylength below which it does influence vernalization rate (h)
    */
    double minDL;
    /**
    * @brief Saturating photoperiod above which final leaf number is not influenced by daylength (h)
    */
    double maxDL;
    /**
    * @brief Maximum temperature for vernalization to occur (°C)
    */
    double maxTvern;
    /**
    * @brief Number of primorida in the apex at emergence (primordia)
    */
    double pNini;
    /**
    * @brief Absolute maximum leaf number (leaf)
    */
    double aMXLFNO;
    /**
    * @brief true if the plant is vernalizable ()
    */
    int isVernalizable;
};
}
}
DECLARE_DYNAMICS(record::Phenology::Vernalizationprogress); // balise specifique VLE